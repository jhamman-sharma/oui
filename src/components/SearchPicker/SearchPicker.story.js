import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';

import Dropdown from '../Dropdown';
import SearchPicker from './index.js';
import Spinner from '../Spinner';
import Icon from 'react-oui-icons';

import { getSearchFunction } from './mock_api';
import Row from '../Layout/Row';

const searchFunction = getSearchFunction();

const stories = storiesOf('Forms|SearchPicker', module);
stories.addDecorator(withKnobs).addDecorator((story) => (
  <div
    id="root-preview"
    style={{
      display: 'flex',
      flex: 1,
      height: '100vh',
    }}>
    {story()}
  </div>
));

stories
  .add('Default', () => {
    return (
      <SearchPicker
        onItemSelected={ action('keyboard selection item') }
        searchFunction={ searchFunction }
        supportedTypes={ ['feature'] }>
        {({
          availableEntities,
          renderInput,
          isLoading,
          searchQuery,
          currentFauxFocusIndex,
          resultsText,
        }) => (
          <Dropdown width={ 600 } fullWidth={ true } renderActivator={ renderInput }>
            <Dropdown.Contents minWidth={ 600 } isLoading={ isLoading }>
              <Dropdown.ListItem ignoreToggle={ true }>
                <span className="micro muted soft--sides">
                  {resultsText.summary}
                </span>
              </Dropdown.ListItem>
              {availableEntities.map((item, index) => (
                <Dropdown.ListItem key={ index }>
                  <Dropdown.BlockLink
                    hasFauxFocus={ currentFauxFocusIndex === index }
                    onClick={ action(`click dropdown block link ${item.name}`) }>
                    <Dropdown.BlockLinkText text={ item.name } />
                    <Dropdown.BlockLinkSecondaryText
                      secondaryText={ item.description }
                    />
                  </Dropdown.BlockLink>
                </Dropdown.ListItem>
              ))}
              {isLoading && <Spinner hasOverlay={ true } />}
            </Dropdown.Contents>
          </Dropdown>
        )}
      </SearchPicker>
    );
  })
  .add('With Create Button', () => {
    return (
      <SearchPicker
        onItemSelected={ action('keyboard selection item') }
        additionalItems={ 1 }
        searchFunction={ searchFunction }
        supportedTypes={ ['feature'] }>
        {({
          availableEntities,
          renderInput,
          isLoading,
          searchQuery,
          currentFauxFocusIndex,
          resultsText,
        }) => (
          <Dropdown width={ 600 } fullWidth={ true } renderActivator={ renderInput }>
            <Dropdown.Contents minWidth={ 600 } isLoading={ isLoading }>
              <Dropdown.ListItem>
                <Dropdown.BlockLink
                  hasFauxFocus={ currentFauxFocusIndex === 0 }
                  onClick={ action('click create dropdown block link') }>
                  <div className="flex flex-align--center">
                    Create new audience
                  </div>
                </Dropdown.BlockLink>
              </Dropdown.ListItem>
              <Dropdown.ListItem ignoreToggle={ true }>
                <span className="micro muted soft--sides">
                  {resultsText.summary}
                </span>
              </Dropdown.ListItem>
              {availableEntities.map((item, index) => (
                <Dropdown.ListItem key={ index }>
                  <Dropdown.BlockLink
                    hasFauxFocus={ currentFauxFocusIndex === index + 1 }
                    onClick={ action(`click dropdown block link ${item.name}`) }>
                    <Dropdown.BlockLinkText text={ item.name } />
                    <Dropdown.BlockLinkSecondaryText
                      secondaryText={ item.description }
                    />
                  </Dropdown.BlockLink>
                </Dropdown.ListItem>
              ))}
              {isLoading && <Spinner hasOverlay={ true } />}
            </Dropdown.Contents>
          </Dropdown>
        )}
      </SearchPicker>
    );
  })
  .add('In a Dropdown', () => {
    return (
      <Dropdown
        buttonContent="Selected Variation Name"
        arrowIcon="down"
        shouldHideChildrenOnClick={ false }
        style="outline">
        {({ handleHideChildren }) => (
          <Dropdown.Contents minWidth={ 600 } direction="right">
            <Dropdown.ListItem>
              <Dropdown.BlockLink
                hasFauxFocus={ false }
                onClick={ action('click create dropdown block link') }>
                <Row verticalAlignment="center" paddedContent="sides">
                  <Icon name="add" />
                  <p className="soft--left flush">Create new variation</p>
                </Row>
              </Dropdown.BlockLink>
            </Dropdown.ListItem>
            <SearchPicker
              onItemSelected={ action('keyboard selection item') }
              additionalItems={ 1 }
              searchFunction={ searchFunction }
              supportedTypes={ ['variation'] }>
              {({
                availableEntities,
                renderInput,
                isLoading,
                searchQuery,
                currentFauxFocusIndex,
                resultsText,
              }) => (
                <React.Fragment>
                  <Dropdown.ListItem>{renderInput()}</Dropdown.ListItem>
                  <Dropdown.ListItem ignoreToggle={ true }>
                    <span className="micro muted soft--sides">
                      {resultsText.summary}
                    </span>
                  </Dropdown.ListItem>
                  {availableEntities.map((item, index) => (
                    <Dropdown.ListItem key={ index }>
                      <Dropdown.BlockLink
                        hasFauxFocus={ currentFauxFocusIndex === index + 1 }
                        onClick={ action(
                          `click dropdown block link ${item.name}`
                        ) }>
                        <Dropdown.BlockLinkText text={ item.name } />
                        <Dropdown.BlockLinkSecondaryText
                          secondaryText={ item.description }
                        />
                      </Dropdown.BlockLink>
                    </Dropdown.ListItem>
                  ))}
                  {isLoading && <Spinner hasOverlay={ true } />}
                </React.Fragment>
              )}
            </SearchPicker>
          </Dropdown.Contents>
        )}
      </Dropdown>
    );
  });
