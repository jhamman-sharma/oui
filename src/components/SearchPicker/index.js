import { debounce, noop } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';

import { keyboardTracker } from '../../utils/accessibility';

import Input from '../Input';

const INITIAL_SEARCH_PAGE_SIZE = 5;
const INPUT_DEBOUNCE_PERIOD = 120;

@keyboardTracker
class SearchPicker extends React.Component {
  static propTypes = {
    /**
     * A render function that is invoked with the following info:
     *   availableEntities {Array<Entity>} - The list of entities available
     *                     based on the current input. If no query term is
     *                     entered, the default search collection is used.
     *
     *   isLoading         {Boolean} - Whether or not a search is in progress.
     *   renderInput       {Function<props>} - A function to render the <Input>
     *   searchQuery       {String} - The current search term
     *   currentFauxFocusIndex {Number} - Index of the item currently selected.
     *   resultsText       {Object} - An object containing search result
     *                     summaries based on the the types provided
     *                     See this.getResultsText
     *
     */
    children: PropTypes.func.isRequired,

    /**
     * Which index in the list is currently selected.
     * via @keyboardTracker
     */
    currentFauxFocusIndex: PropTypes.number.isRequired,

    /**
     * A handler for the input element event onKeyDown
     * via @keyboardTracker
     */
    handleKeyDown: PropTypes.func.isRequired,

    /**
     * The element ref to use for capturing keyboard input.
     * via @keyboardTracker
     */
    keyboardRef: PropTypes.shape({ current: PropTypes.elementType }),

    /**
     * Handler to call when an item is selected via the keyboard.
     */
    onItemSelected: PropTypes.func.isRequired,

    /**
     * The function to invoke to execute an authenticated public API search.
     * Different consumers will have different methods of authenticating.
     *
     * @param {Object} requestConfig
     * @return {Promise}
     */
    searchFunction: PropTypes.func.isRequired,

    /**
     * Additional options to pass to the search function when a search is
     * executed. Type and query will be overridden by this component.
     */
    searchOptions: PropTypes.object,

    /**
     * An optional list of ids which should be considered "selected".
     * If provided, these ids will be filtered out of the list.
     */
    selectedEntityIds: PropTypes.array,

    /**
     * Sets the number of items for the list index tracker.
     * via @keyboardTracker
     */
    setItemCount: PropTypes.func,

    /**
     * Sets the item select callback fn for the list index tracker.
     * via @keyboardTracker
     */
    setOnItemSelect: PropTypes.func,

    /* The entity types we should query search for. */
    supportedTypes: PropTypes.array.isRequired,
  };

  static defaultProps = {
    searchOptions: {},
    selectedEntityIds: [],
    setItemCount: noop,
    setOnItemSelect: noop,
  };

  /**
   * Notify keyboardTracker what to do when ENTER is pressed on an item
   * and kickoff the initial search to populate the default result set.
   */
  componentDidMount = () => {
    const { searchFunction, searchOptions, setOnItemSelect, supportedTypes } = this.props;
    setOnItemSelect(() => this.handleItemSelected);

    searchFunction({
      ...searchOptions,
      types: supportedTypes,
      query: '',
      per_page: INITIAL_SEARCH_PAGE_SIZE,
    }).then(results => {
      // Make this Promise catch if the results are not an array
      this.verifyResultOrThrow(results);
      this.setState({
        defaultResults: results,
      });
    })
      .finally(() => {
        const { searchQuery } = this.state;
        // Always set loading false as long as searchQuery is empty
        if (!searchQuery) {
          this.setState({
            isLoading: false,
          });
        }
      });
  };

  /**
   * Whenever the component updates, notify the keyboardTracker
   * of how many items are currently available. (this also resets
   * the faux focus index to 0)
   */
  componentDidUpdate = () => {
    const items = this.getAvailableEntities();
    this.props.setItemCount(items.length);
  };

  state = {
    currentSearch: null,
    defaultResults: [],
    isLoading: true,
    results: null,
    searchQuery: '',
  };

  handleItemSelected = index => {
    const { onItemSelected } = this.props;
    const { searchQuery } = this.state;
    onItemSelected(index, this.getAvailableEntities(), searchQuery);
  };

  /**
   * Handles changes to the search term by calling the debounced change handler.
   * If the term is an empty string, short circuit and reset the results.
   * @param {React.SyntheticEvent} event - The onInput event
   */
  handleOnInput = event => {
    const { value } = event.target;
    this.setState({
      searchQuery: value,
    });

    if (value) {
      this.handleOnInputDebounced(value);
    } else {
      const { currentSearch } = this.state;
      if (currentSearch) {
        currentSearch.reject('Search interrupted by empty search query.');
      }
      this.setState({
        isLoading: false,
        results: null,
        currentSearch: null,
      });
    }
  };

  /**
   * Execute the search function in a way that is cancellable by storing
   * the Promise resolution methods in state.
   * This enables subsequent invocations to cancel an in progress
   * search in favor of a new one.
   * @param {string} query - The term to search for.
   * @returns {Promise<Array>}
   */
  executeNewSearch = query =>
    new Promise((resolve, reject) => {
      const { searchFunction, searchOptions, supportedTypes } = this.props;
      const { currentSearch } = this.state;
      if (currentSearch) {
        currentSearch.reject(`Search interrupted by new search query: ${query}`);
      }
      this.setState({
        currentSearch: {
          resolve,
          reject,
        },
        isLoading: true,
      });
      searchFunction({
        ...searchOptions,
        types: supportedTypes,
        query: query.toLowerCase(),
      })
        .then(resolve, reject);
    });

  /**
   * A debounced method to invoke the search and store the result in state.
   * @param {string} query - The term to search for.
   */
  handleOnInputDebounced = debounce(query => {
    this.executeNewSearch(query).then(results => {
      const { searchQuery } = this.state;
      this.setState({
        isLoading: false,
        currentSearch: null,
      });
      // Ensure the query wasn't killed while waiting for the result.
      if (searchQuery) {
        this.setState({
          results,
        });
      }
    }, e => { /** Swallow rejections */});
  }, !!process.env.JEST_WORKER_ID ? 0 : INPUT_DEBOUNCE_PERIOD);

  /**
   * Filter the current result set to exclude selected entities.
   * @returns {Array<Object>}
   */
  getAvailableEntities = () => {
    const { selectedEntityIds } = this.props;
    return this.getResultSet().filter(result => !selectedEntityIds.includes(result.id));
  };

  /**
   * Helper method to return the current result set,
   * falling back to the default.
   * @returns {Array<Object>}
   */
  getResultSet = () => {
    const { results, defaultResults } = this.state;
    return results || defaultResults;
  };

  /**
   * Compute a human readable result text.
   * @returns {Object}
   */
  getResultsText = () => {
    const { supportedTypes } = this.props;
    const { isLoading, searchQuery } = this.state;
    const resultCount = this.getAvailableEntities().length;
    const summaryNoun = supportedTypes.length > 1 ? 'entities' : `${supportedTypes[0]}s`;
    let summary = `Recently created ${summaryNoun}`;

    if (isLoading) {
      summary = `Searching for "${summaryNoun}" matching "${searchQuery}"`;
    } else if (searchQuery) {
      if (resultCount) {
        summary = `Found ${resultCount} ${summaryNoun} matching "${searchQuery}"`;
      } else {
        summary = `No ${summaryNoun} found matching "${searchQuery}"`;
      }
    }

    return {
      summary,
    };
  };

  /**
   * @description Provided an API response, an error will be thrown if the
   *  provided value is not an Array. This is used within a Promise and
   *  will only cause the Promise to catch instead of this component.
   * @param {*} results - The response of the provided searchFunction, which
   *  is expected to be an Array.
   */
  verifyResultOrThrow = results => {
    if (!Array.isArray(results)) {
      throw new Error(
        'SearchPicker props.searchFunction resolved with a value that is not an Array.'
      );
    }
  };

  /**
   * A render method passed to children so that they can render the search
   * input component wherever they like.
   * @param {Object} props - Any additional props to pass to the <Input>.
   * @returns {React.Component}
   */
  renderInput = (props = {}) => {
    const { handleKeyDown, supportedTypes } = this.props;
    const { searchQuery } = this.state;
    const { handleOnInput } = this;
    const placeholderNoun = supportedTypes.map(i => `${i}s`).join(',');
    return (
      <Input
        isFilter={ true }
        onChange={ handleOnInput }
        placeholder={ `Search for ${placeholderNoun}` }
        testSection={ `filter-${placeholderNoun}-input` }
        type="search"
        value={ searchQuery }
        onKeyDown={ handleKeyDown }
        { ...props }
      />
    );
  };

  /**
   * A simple render method that passes relevant state and controls to the
   * children.
   * @returns {React.Component}
   */
  render() {
    const { children, currentFauxFocusIndex } = this.props;
    const { searchQuery, isLoading } = this.state;
    const { renderInput, handleOnInputDebounced } = this;
    const availableEntities = this.getAvailableEntities();
    return (
      children({
        availableEntities,
        isLoading,
        renderInput,
        searchQuery,
        makeSearchRequest: handleOnInputDebounced,
        currentFauxFocusIndex,
        resultsText: this.getResultsText(),
      })
    );
  }
}

export default SearchPicker;
