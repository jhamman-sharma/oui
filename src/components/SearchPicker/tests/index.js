import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';

import { getSearchFunction } from '../mock_api';
import MOCK_DATA from '../mock_api/mock_data';

import SearchPicker from '../';

const MOCK_API_DELAY = 100;

const searchFunction = getSearchFunction(MOCK_API_DELAY);

const getDelayedPromise = (delay = 0) => new Promise(r => setTimeout(r, delay));


describe('components/SearchPicker', () => {
  let component;
  let renderedData;
  let onItemSelectedStub;

  describe('default case', () => {
    beforeEach(() => {
      onItemSelectedStub = jest.fn();
      component = mount(
        <SearchPicker
          onItemSelected={ onItemSelectedStub }
          searchFunction={ searchFunction }
          supportedTypes={ ['feature'] }>
          {(data) => {
            renderedData = data;
            return (
              <div>
                {data.renderInput()}
                <p>A list of entities</p>
              </div>
            );
          }}
        </SearchPicker>
      );
    });

    it('should render component same as snapshot', () => {
      expect(mountToJson(component)).toMatchSnapshot();
    });

    it('should pass a default list of items on creation', () => {
      return getDelayedPromise().then(() => {
        component.find('input').simulate('click');
      })
        .then(() => getDelayedPromise(MOCK_API_DELAY))
        .then(() => {
          component.update();
          expect(renderedData.availableEntities).toHaveLength(5);
        });
    });

    it('should update the list when a search term is entered', () => {
      return getDelayedPromise().then(() => {
        component.find('input').simulate('mouseenter');
        component.find('input').simulate('click');
        component.find('input').simulate('change', { target: { value: '123' }});
      })
        // Account for our mock API delay and the debounce delay
        .then(() => getDelayedPromise(MOCK_API_DELAY + 10))
        .then(() => {
          component.update();
          expect(renderedData.availableEntities).toHaveLength(1);
          expect(renderedData.availableEntities[0].name).toContain('My feature 123');
        });
    });

    it('should pass isLoading, searchQuery, and availableEntities correctly', () => {
      expect(renderedData.isLoading).toBe(true);
      return getDelayedPromise(MOCK_API_DELAY).then(() => {
        expect(renderedData.searchQuery).toBe('');
        expect(renderedData.isLoading).toBe(false);
        component.find('input').simulate('mouseenter');
        component.find('input').simulate('click');
        component.find('input').simulate('change', { target: { value: '123' }});
      })
        // Wait for debounce
        .then(() => getDelayedPromise(10))
        .then(() => {
          expect(renderedData.searchQuery).toBe('123');
          expect(renderedData.isLoading).toBe(true);
        })
        // Account for our mock API delay and the debounce delay
        .then(() => getDelayedPromise(MOCK_API_DELAY + 10))
        .then(() => {
          component.update();
          expect(renderedData.isLoading).toBe(false);
          expect(renderedData.availableEntities).toEqual([MOCK_DATA[0]]);
        });
    });
    it('should wait to show results until all queries are completed', function() {
      expect(renderedData.isLoading).toBe(true);
      return getDelayedPromise(MOCK_API_DELAY).then(() => {
        expect(renderedData.resultsText).toEqual({summary: 'Recently created features'});
        component.find('input').simulate('mouseenter');
        component.find('input').simulate('click');
        component.find('input').simulate('change', { target: { value: '123' }});
      })
        // Wait for debounce
        .then(() => getDelayedPromise(10))
        .then(() => {
          expect(renderedData.searchQuery).toBe('123');
          expect(renderedData.isLoading).toBe(true);
          component.update();
          expect(renderedData.resultsText).toEqual({summary: 'Searching for "features" matching "123"'});
          component.find('input').simulate('mouseenter');
          component.find('input').simulate('click');
          component.find('input').simulate('change', { target: { value: '202' }});
        })
        .then(() => getDelayedPromise(MOCK_API_DELAY))
        .then(() => {
          component.update();
          expect(renderedData.resultsText).toEqual({summary: 'Searching for "features" matching "202"'});
        })
        // Account for our mock API delay and the debounce delay
        .then(() => getDelayedPromise(MOCK_API_DELAY + 10))
        .then(() => {
          component.update();
          expect(renderedData.isLoading).toBe(false);
          expect(renderedData.resultsText).toEqual({summary: 'Found 1 features matching "202"'});
        });
    });
    it('should render the number of available entities in the results text', function() {
      // We need to pass in selected entities to properly test this
      component.setProps({ selectedEntityIds: [202, 205, 207] });
      component.update();
      component.find('input').simulate('mouseenter');
      component.find('input').simulate('click');
      component.find('input').simulate('change', { target: { value: 'audience 20' }});
      return getDelayedPromise(MOCK_API_DELAY + 10).then(() => {
        component.update();
        expect(renderedData.resultsText).toEqual({summary: 'Found 4 features matching "audience 20"'});
      });
    });
  });

  describe('searchFunction responses', () => {
    it('should update state when an Array is returned from searchFunction', () => {
      onItemSelectedStub = jest.fn();
      const allGoodSearchFn = jest.fn(() => new Promise(resolve => resolve([
        {
          id: 123,
          name: 'My feature 123',
          type: 'feature',
          description: 'A very cool feature',
        },
      ])));
      component = mount(
        <SearchPicker
          onItemSelected={ onItemSelectedStub }
          searchFunction={ allGoodSearchFn }
          supportedTypes={ ['feature'] }>
          {(data) => {
            renderedData = data;
            return (
              <div>
                {data.renderInput()}
                <p>A list of entities</p>
              </div>
            );
          }}
        </SearchPicker>
      );
      return allGoodSearchFn().then(() => {
        const defaultResultsState = component.children().instance().state.defaultResults;
        expect(defaultResultsState).toEqual([
          {
            id: 123,
            name: 'My feature 123',
            type: 'feature',
            description: 'A very cool feature',
          },
        ]);
      });
    });
    it('should not update state when an object is returned from searchFunction', () => {
      onItemSelectedStub = jest.fn();
      const notGoodSearchFn = jest.fn((() => new Promise(resolve => resolve({}))));
      component = mount(
        <SearchPicker
          onItemSelected={ onItemSelectedStub }
          searchFunction={ notGoodSearchFn }
          supportedTypes={ ['feature'] }>
          {(data) => {
            renderedData = data;
            return (
              <div>
                {data.renderInput()}
                <p>A list of entities</p>
              </div>
            );
          }}
        </SearchPicker>
      );
      return notGoodSearchFn().then(() => {
        const defaultResultsState = component.children().instance().state.defaultResults;
        expect(defaultResultsState).toEqual([]);
      });
    });
    it('should not update state when undefined is returned from searchFunction', () => {
      onItemSelectedStub = jest.fn();
      const notGoodSearchFn = jest.fn((() => new Promise(resolve => resolve(undefined))));
      component = mount(
        <SearchPicker
          onItemSelected={ onItemSelectedStub }
          searchFunction={ notGoodSearchFn }
          supportedTypes={ ['feature'] }>
          {(data) => {
            renderedData = data;
            return (
              <div>
                {data.renderInput()}
                <p>A list of entities</p>
              </div>
            );
          }}
        </SearchPicker>
      );
      return notGoodSearchFn().then(() => {
        const defaultResultsState = component.children().instance().state.defaultResults;
        expect(defaultResultsState).toEqual([]);
      });
    });
  });
});
