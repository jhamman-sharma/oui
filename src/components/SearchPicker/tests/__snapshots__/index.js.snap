// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SearchPicker default case should render component same as snapshot 1`] = `
<withkeyboardTracker(undefined)
  additionalItems={0}
  onItemSelected={[MockFunction]}
  searchFunction={[Function]}
  supportedTypes={
    Array [
      "feature",
    ]
  }
>
  <SearchPicker
    currentFauxFocusIndex={0}
    handleKeyDown={[Function]}
    onItemSelected={[MockFunction]}
    searchFunction={[Function]}
    searchOptions={Object {}}
    selectedEntityIds={Array []}
    setItemCount={[Function]}
    setOnItemSelect={[Function]}
    supportedTypes={
      Array [
        "feature",
      ]
    }
  >
    <div>
      <[object Object]
        hasSpellCheck={true}
        isFilter={true}
        isRequired={false}
        note={null}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search for features"
        testSection="filter-features-input"
        type="search"
        value=""
      >
        <input
          className="oui-text-input oui-text-input--search"
          data-oui-component={true}
          data-test-section="filter-features-input"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Search for features"
          required={false}
          spellCheck={true}
          type="search"
          value=""
        />
      </[object Object]>
      <p>
        A list of entities
      </p>
    </div>
  </SearchPicker>
</withkeyboardTracker(undefined)>
`;
